### DISCUSSION TITLE
Function with declared return type "str" must return value on all code paths

### DISCUSSION THREAD

**Discussion by aaaaahaaaaa** [microsoft/pyright#6404] [General]
```python
def foo() -> str:
    for i in range(10):
        return ""
```

Why does the code above raise the following error?
```
Function with declared return type "str" must return value on all code paths
  Type "None" cannot be assigned to type "str"
```

**Comment by erictraut** [comment]
It's a type violation because type checkers cannot determine statically how many times the `for` loop will execute. The type of `range(10)` is `Iterable[int]` which means that it can return zero or more items. If it returns zero, the code will fall through, and the function will return `None`. Of course, we happen to know that `range(10)` will never return zero items, but that's not visible to a type checker, which bases its knowledge on the type information it's provided. There's currently no mechanism in the type system for an `Iterable` to be flagged as "always returning at least one item".

**Reply by aaaaahaaaaa** [reply]
I see. Thank you for the clear explanation.
